enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/outcome/Outcome.java

```diff
@@ -20,54 +20,144 @@
 package org.neo4j.coreedge.raft.outcome;
 
 import java.io.Serializable;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.HashSet;
 import java.util.Set;
 
 import org.neo4j.coreedge.raft.RaftMessages;
 import org.neo4j.coreedge.raft.roles.Role;
 import org.neo4j.coreedge.raft.state.FollowerStates;
+import org.neo4j.coreedge.raft.state.ReadableRaftState;
 
+/**
+ * Holds the outcome of a RAFT role's handling of a message. The role handling logic is stateless
+ * and responds to RAFT messages in the context of a supplied state. The outcome is later consumed
+ * to update the state and do operations embedded as commands within the outcome.
+ *
+ * A state update could be to change role, change term, etc.
+ * A command could be to append to the RAFT log, tell the log shipper that there was a mismatch, etc.
+ */
 public class Outcome<MEMBER> implements Serializable
 {
-    public final Role newRole;
-    public final long newTerm;
-    public final MEMBER leader;
-    public final long leaderCommit;
-    public final MEMBER votedFor;
-    public final Set<MEMBER> votesForMe;
-    public final long lastLogIndexBeforeWeBecameLeader;
-    public final FollowerStates<MEMBER> followerStates;
-    public final boolean renewElectionTimeout;
-    public final Iterable<LogCommand> logCommands;
-    public final Iterable<ShipCommand> shipCommands;
-    public final Collection<RaftMessages.Directed<MEMBER>> outgoingMessages;
-
-    public Outcome(Role newRole, long newTerm, MEMBER leader, long leaderCommit, MEMBER votedFor,
-                   Set<MEMBER> votesForMe, long lastLogIndexBeforeWeBecameLeader,
-                   FollowerStates<MEMBER> followerStates, boolean renewElectionTimeout,
-                   Iterable<LogCommand> logCommands, Collection<RaftMessages.Directed<MEMBER>> outgoingMessages,
-                   Iterable<ShipCommand> shipCommands )
+    /* Common */
+    public Role newRole;
+
+    public long term;
+    public MEMBER leader;
+
+    public long leaderCommit;
+
+    public ArrayList<LogCommand> logCommands = new ArrayList<>();
+    public ArrayList<RaftMessages.Directed<MEMBER>> outgoingMessages = new ArrayList<>();
+
+    /* Follower */
+    public MEMBER votedFor;
+    public boolean renewElectionTimeout;
+
+    /* Candidate */
+    public HashSet<MEMBER> votesForMe;
+    public long lastLogIndexBeforeWeBecameLeader;
+
+    /* Leader */
+    public FollowerStates<MEMBER> followerStates;
+    public ArrayList<ShipCommand> shipCommands = new ArrayList<>();
```

Comment:
@martinfurmanski Having public non-final fields is a bad design in principle imho. Here it is even more confusing what you are trying to achieve since you have non-final public fields and setters.